### PR TITLE
TASK: Allow sizes and multiplier for container

### PR DESCRIPTION
**Pull request by extcode**
* Daniel Siepmann couldn't resist to miss use this review to refactor the architecture.

Related: #22

### CODE REVIEW

File: Classes/Sizes/Container.php [hunk 1]
```diff
@@ -58,20 +59,14 @@ public function getActiveColumn(): Column
...
+            $this->contentType,
+        ]);
 
-    /**
-     * @return int[]
-     */
-    public function getSizes(): array
-    {
-        return $this->getActiveColumn()->getSizes();
+        $this->scalingConfiguration = $this->readConfigurationByPath($configurationPath);
```
Comment (justusmoroni) [inline Classes/Sizes/Container.php:69]:
```suggestion
        $this->scalingConfiguration = $this->readConfigurationByPath(implode('.', [
            'container',
            $this->contentType,
        ]));
```

File: Classes/Sizes/Container.php [hunk 2]
```diff
@@ -58,20 +59,14 @@ public function getActiveColumn(): Column
...
-     */
-    public function getMultiplier(): array
+    public function readConfiguration(): void
     {
-        return $this->getActiveColumn()->getMultiplier();
-    }
+        $configurationPath = implode('.', [
+            'container',
+            $this->contentType,
+        ]);
```
Comment (justusmoroni) [inline Classes/Sizes/Container.php:67]:
```suggestion
```

File: Classes/Sizes/Rootline.php [hunk 3]
```diff
@@ -144,21 +170,17 @@ private function calculateSizes(): void
...
 
         foreach ($this->rootline as $contentElement) {
-            if ($contentElement instanceof ContentElementInterface) {
-                $sizes = $contentElement->getSizes();
-                if (!empty($sizes)) {
-                    break;
-                }
-                $multiplier[] = $contentElement->getMultiplier();
+            $sizes = [];
+            [$sizes, $multiplier] = $this->getSizesAndMultiplierFromContentElement($contentElement, $sizes, $multiplier);
```
Comment (justusmoroni) [inline Classes/Sizes/Rootline.php:176]:
```suggestion
            [$sizes, $multiplier] = $this->getSizesAndMultiplierFromContentElement($contentElement, $multiplier);
```

Comment (extcode) [reply Classes/Sizes/Rootline.php]: Sollte erledigt sein.

File: Classes/Sizes/Rootline.php [hunk 4]
```diff
@@ -59,6 +59,32 @@ public function getFinalSizes(): array
         return $this->finalSizes;
     }
 
+    public function getSizesAndMultiplierFromContentElement(
+        mixed $contentElement,
+        array $sizes,
+        array $multiplier
+    ): array {
+        if ($contentElement instanceof ContentElementInterface) {
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Sizes/Rootline.php]: Sollte erledigt sein.

File: Classes/Sizes/Rootline.php [hunk 5]
```diff
@@ -59,6 +59,32 @@ public function getFinalSizes(): array
         return $this->finalSizes;
     }
 
+    public function getSizesAndMultiplierFromContentElement(
+        mixed $contentElement,
+        array $sizes,
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Sizes/Rootline.php]: Sollte erledigt sein.

File: Classes/Sizes/Container.php (hunk 2 above)
Comment (extcode) [reply Classes/Sizes/Container.php]: Sollte erledigt sein.

File: Classes/Sizes/Container.php (hunk 1 above)
Comment (extcode) [reply Classes/Sizes/Container.php]: Sollte erledigt sein.

File: Classes/Sizes/Rootline.php [hunk 6]
```diff
@@ -38,145 +36,111 @@ final class Rootline
...
     private string $fieldName;
 
-    public function __construct(array $data, string $fieldName)
-    {
+    private string $backendLayoutIdentifier;
+
+    public function __construct(
+        array $data,
+        string $fieldName
+    ) {
```
Comment (DanielSiepmann) [inline Classes/Sizes/Rootline.php:50]: Please start with turning this class in a proper factory.
The `getFinalSizes()` method should be the actual factory method. The data passed to construct should be passed to the method.

That would make this class stateless. Instead the method would act as factory and build the actual array.
That way the rootline can be injected into consumer classes. And the factory can define its own dependencies via constructor.
I'd also state that this is a factory by renaming the class.

File: Classes/Sizes/Rootline.php [hunk 7]
```diff
@@ -38,145 +36,111 @@ final class Rootline
...
+        $sizes = $this->finalSizes;
+
+        foreach ($sizes as $sizeName => &$size) {
+            $size = ceil($size);
+        }
+
+        return $sizes;
+    }
+
+    public function getSizesAndMultiplierFromContentElement(
```
Comment (DanielSiepmann) [inline Classes/Sizes/Rootline.php:73]: I couldn't find any usage of this method. Do we still need the method?

Comment (extcode) [reply Classes/Sizes/Rootline.php]: Sollte erledigt sein.

File: Classes/Sizes/AbstractRootlineElement.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use Codappix\ResponsiveImages\Configuration\ConfigurationManager;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+class AbstractRootlineElement
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Sizes/AbstractRootlineElement.php]: Should be fixed.

File: Classes/Sizes/AbstractContentElement.php [hunk 9]
```diff
@@ -23,25 +23,20 @@
...
-    protected ConfigurationManager $configurationManager;
-
     protected int $colPos;
 
     protected string $contentType;
 
-    protected array $data;
-
-    protected ContentElementInterface $parent;
+    protected RootlineElementInterface $parent;
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Sizes/AbstractContentElement.php]: Should be fixed.

File: Classes/Sizes/AbstractRootlineElement.php [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    protected RootlineElementInterface $parent;
+
+    protected ScalingConfiguration $scalingConfiguration;
+
+    public function __construct()
+    {
+        $this->configurationManager = GeneralUtility::makeInstance(ConfigurationManager::class);
+    }
+
+    public function getParent(): ?RootlineElementInterface
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Sizes/AbstractRootlineElement.php:42]: I allow null for this property and initialize $parent with null.

File: Classes/Domain/Repository/ContainerRepository.php [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+class ContainerRepository
+{
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     * @throws Exception
+     */
+    public function findByIdentifier(int $identifier): array
+    {
+        /** @var QueryBuilder $queryBuilder */
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Domain/Repository/ContainerRepository.php]: I remove the comment with replacing the code with DI.

File: Classes/Domain/Repository/ContainerRepository.php [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable('tt_content');
+        $rawData = $queryBuilder
+            ->select('*')
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Domain/Repository/ContainerRepository.php]: I limit the select to required fields.

File: Classes/Domain/Repository/ContainerRepository.php [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+class ContainerRepository
+{
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     * @throws Exception
+     */
+    public function findByIdentifier(int $identifier): array
+    {
+        /** @var QueryBuilder $queryBuilder */
+        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable('tt_content');
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Domain/Repository/ContainerRepository.php]: The DI is used here.

File: Classes/Sizes/AbstractRootlineElement.php [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public function setParent(RootlineElementInterface $rootlineElement): void
+    {
+        $this->parent = $rootlineElement;
+    }
+
+    public function getScalingConfiguration(): ScalingConfiguration
+    {
+        return $this->scalingConfiguration;
```
[the remark this answers is not in the document]
Comment (extcode) [reply Classes/Sizes/AbstractRootlineElement.php:54]: Rootline is an Factory now.

File: Classes/Sizes/Rootline.php (hunk 6 above)
Comment (extcode) [reply Classes/Sizes/Rootline.php]: I think this is already fixed.

**Review by DanielSiepmann — APPROVED** [review]

**Review by justusmoroni — APPROVED** [review]

**Pull request merged**